src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -949,8 +949,28 @@ private void InitializePackagesFolderInfo(string packagesFolderPath, bool exclud
                     // we are targeting a particular package - there is no need therefore to alter other aspects of the project
                     // specifically an unrelated package may have been force removed in which case we should be happy to leave things that way
 
+                    var newPackagesToUninstallUpdated = new List<PackageIdentity>();
+                    foreach (var oldInstalledPackage in oldListOfInstalledPackages)
```

Comment:
I think most of this can be simplified into:

var newTargetIds = new HashSet<string>(oldListOfInstalledPackages.Where(oldPackage => newListOfInstalledPackages.Any(newPackage => StringComparer.OrdinalIgnoreCase.Equals(oldPackage.Id, newPackage.Id) && !oldPackage.Version.Equals(newPackage.Version)).Select(package => package.Id), StringComparer.OrdinalIgnoreCase)